### ISSUE TITLE
Error when helm installing ./hlf-kub on raspberry Pi-4

### ISSUE THREAD

**Issue by weishancc** [hyfen-nl/PIVT#87]
Hi Everyone,

I'm trying to run K8s cluster on RPi 4 (with 1 worker node on AMD PC / 2 worker nodes on ARM v8 RPi-4). 
Everything go smoothly until command:
`helm install ./hlf-kube --name hlf-kube -f samples/simple/network.yaml -f samples/simple/crypto-config.yaml`

This is the error message log of tiller-deploy (tiller ARM version from [jessestuart/tiller](https://github.com/jessestuart/tiller-multiarch)), seems it cannot connect to api-server: 

 `[storage] 2020/09/21 06:58:13 getting release history for "hlf-kube"`
 `[storage/driver] 2020/09/21 06:58:43 query: failed to query with labels: Get https://10.96.0.1:443/api/v1/namespaces/kube-system/configmaps?labelSelector=NAME%3Dhlf-kube%2COWNER%3DTILLER: dial tcp 10.96.0.1:443: i/o timeout
 [tiller] 2020/09/21 06:59:13 failed install prepare step: Get https://10.96.0.1:443/version?timeout=32s: dial tcp 10.96.0.1:443: i/o timeout`

Before that, tiller can be successfully assigned to one of worker node 
```console
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  25m   default-scheduler  Successfully assigned kube-system/tiller-deploy-568ff59df6-tv6wd to org2
  Normal  Pulled     25m   kubelet, org2      Container image "jessestuart/tiller" already present on machine
  Normal  Created    25m   kubelet, org2      Created container tiller
  Normal  Started    25m   kubelet, org2      Started container tiller
```

Can anyone help with me? Thanks!

**Comment by raftAtGit** [comment]
hehe, HL Fabric on Raspberry Pi ❤️ 

dont know what is going on there, but dont think the error you got is related to this repo. anyway:

1. this might be related to incorrect installation of tiller, especially if RBAC is enabled
2. can be a temporary thing, try deleting `kube-proxy` pods in `kube-system` namespace. I've seen many times timeout errors are resolved after restarting `kube-proxy` pods.

in either case, make sure `helm list` command returns a result without error before proceeding to installation of Fabric

**Comment by weishancc** [comment]
> hehe, HL Fabric on Raspberry Pi ❤️
> 
> dont know what is going on there, but dont think the error you got is related to this repo. anyway:
> 
> 1. this might be related to incorrect installation of tiller, especially if RBAC is enabled
> 2. can be a temporary thing, try deleting `kube-proxy` pods in `kube-system` namespace. I've seen many times timeout errors are resolved after restarting `kube-proxy` pods.
> 
> in either case, make sure `helm list` command returns a result without error before proceeding to installation of Fabric

Thanks for the direction.  
`helm list` has the same error so, I'll figure out !

**Comment by weishancc** [comment]
### Hi, Raft : )

I've successfully build hlf on pi4. However, when trying to create channel and join peer with argo, i always got the hlf-channels- status `Pending` as the screenshot below.

![image](https://user-images.githubusercontent.com/44445847/95191166-ed5a3180-0802-11eb-820d-17264776c62e.png)
![image](https://user-images.githubusercontent.com/44445847/95191252-0f53b400-0803-11eb-95b6-eb220dfe7167.png)

What i'm doing is changing fabric image in `channel-workflow.yaml` to arm version. And argo itself seems work well. Do u have any suggestion with it? Or something else need to be modified in channel workflow ? Thank you!

**argo pods**
![image](https://user-images.githubusercontent.com/44445847/95191382-475af700-0803-11eb-9132-6a26a1f837c8.png)

**Comment by raftAtGit** [comment]
`kubectl describe` command can give more information why workflow stuck at pending state:
```
kubectl describe workflow hlf-channels-xxx
```
also check argo workflow controller pod's log. deleting workflow controller pod (forcing it to restart) may also help.

> What i'm doing is changing fabric image in channel-workflow.yaml to arm version

not all the images have arm version, in particular `raft/hl-fabric-tools:1.4.3` one, which is our home built image used for declarative part of the flow

**Comment by weishancc** [comment]
I replaced [fabric-tools](https://hub.docker.com/layers/ptunstad/fabric-tools/arm64-1.4.1/images/sha256-b2aa4499b4445c253159928a3917236d704112114b8a96f37a2f13b502e6d902?context=explore) to arm version as well as commented out every check comment (checkjoined, checkanchors, checkexists) as initial.

Just have no idea after seeing workflow's log. Only if the whole workflow yaml correct can i run this workflow successful? I wonder how to see the status of every step (i.e. create channel ,fetch block, join peer, update anchor) and which step the error occurred in this log.

[workflow_log.txt](https://github.com/hyfen-nl/PIVT/files/5337867/workflow_log.txt)
[controller_log.txt](https://github.com/hyfen-nl/PIVT/files/5337835/controller_log.txt)

**Comment by raftAtGit** [comment]
if workflow yaml is not correct, argo command will not accept it.
once the workflow started, you can see logs of each step by `argo logs <pod name>` or `kubectl logs <pod name> -c main`

did you try deleting argo controller pod?

before continuing, I would suggest make sure you can run any sample argo workflow.

instead of commenting out `checkexists` etc, you can use [release/0.7](https://github.com/APGGroeiFabriek/PIVT/tree/release/0.7) which is not declarative and only uses official Fabric images.

**Comment by weishancc** [comment]
Yes i've tried to delete argo controller pod and run sample argo workflow. 
After that i just added namespace `-n argo` in argo submit command then the workflow began to run : )
![image](https://user-images.githubusercontent.com/44445847/95705273-fa08da80-0c85-11eb-88dc-e735bb04e658.png)

Now the workflow stuck when creating channel (same error occurred in **master** and **release/0.7**) and here's the log: 

**k describe pod hlf-channels-v4cdw-516465887 -n argo**
**(several secrets not found)**
```console
Name:           hlf-channels-v4cdw-516465887
Namespace:      argo
Priority:       0
Node:           org2/10.1.1.100
Start Time:     Mon, 12 Oct 2020 05:16:02 +0100
Labels:         workflows.argoproj.io/completed=false
                workflows.argoproj.io/workflow=hlf-channels-v4cdw
Annotations:    workflows.argoproj.io/node-name: hlf-channels-v4cdw[0].channel--common[0].create-channel--common[0].create-channel--common--karga--peer9(0)
                workflows.argoproj.io/template:
                  {"name":"create-channel--common--karga--peer9","arguments":{},"inputs":{},"outputs":{},"metadata":{},"container":{"name":"","image":"ptuns...
Status:         Pending
IP:             
IPs:            <none>
Controlled By:  Workflow/hlf-channels-v4cdw
Containers:
  wait:
    Container ID:  
    Image:         argoproj/argoexec:v2.11.1
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_POD_NAME:  hlf-channels-v4cdw-516465887 (v1:metadata.name)
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /argo/secret/my-minio-cred from my-minio-cred (ro)
      /mainctrfs/etc/hyperledger/fabric/msp from peer-karga-peer9-admin-msp (rw)
      /mainctrfs/etc/hyperledger/fabric/tls from peer-karga-peer9-tls (rw)
      /mainctrfs/hlf_config/channel from hlf-channels (rw)
      /mainctrfs/hlf_config/orderer-tlsca from orderer-tlsca (rw)
      /var/run/docker.sock from docker-sock (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-prx47 (ro)
  main:
    Container ID:  
    Image:         ptunstad/fabric-peer:arm64-1.4.1
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
      -c
      peer channel create -o orderer0.groeifabriek.nl:7050 -c common -f /hlf_config/channel/common.tx --tls --cafile /hlf_config/orderer-tlsca/tlscacert.pem 
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:
      FABRIC_LOGGING_SPEC:          debug
      CORE_PEER_LOCALMSPID:         KargaMSP
      CORE_PEER_TLS_CERT_FILE:      /etc/hyperledger/fabric/tls/server.crt
      CORE_PEER_TLS_KEY_FILE:       /etc/hyperledger/fabric/tls/server.key
      CORE_PEER_TLS_ROOTCERT_FILE:  /etc/hyperledger/fabric/tls/ca.crt
    Mounts:
      /etc/hyperledger/fabric/msp/ from peer-karga-peer9-admin-msp (rw)
      /etc/hyperledger/fabric/tls/ from peer-karga-peer9-tls (rw)
      /hlf_config/channel/ from hlf-channels (rw)
      /hlf_config/orderer-tlsca from orderer-tlsca (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-prx47 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  docker-sock:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/docker.sock
    HostPathType:  Socket
  orderer-tlsca:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hlf-orderer--groeifabriek-tlsca
    Optional:    false
  peer-karga-peer9-tls:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hlf-peer--karga--peer9-tls
    Optional:    false
  peer-karga-peer9-admin-msp:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hlf-peer--karga--peer9-admin-msp
    Optional:    false
  hlf-channels:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hlf-channels
    Optional:    false
  my-minio-cred:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  my-minio-cred
    Optional:    false
  default-token-prx47:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-prx47
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason       Age                   From               Message
  ----     ------       ----                  ----               -------
  Normal   Scheduled    34m                   default-scheduler  Successfully assigned argo/hlf-channels-v4cdw-516465887 to org2
  Warning  FailedMount  34m (x6 over 34m)     kubelet, org2      MountVolume.SetUp failed for volume "peer-karga-peer9-admin-msp" : secret "hlf-peer--karga--peer9-admin-msp" not found
  Warning  FailedMount  34m (x6 over 34m)     kubelet, org2      MountVolume.SetUp failed for volume "peer-karga-peer9-tls" : secret "hlf-peer--karga--peer9-tls" not found
  Warning  FailedMount  19m                   kubelet, org2      (combined from similar events): Unable to attach or mount volumes: unmounted volumes=[orderer-tlsca peer-karga-peer9-tls peer-karga-peer9-admin-msp hlf-channels], unattached volumes=[my-minio-cred orderer-tlsca peer-karga-peer9-tls peer-karga-peer9-admin-msp hlf-channels default-token-prx47 podmetadata docker-sock]: timed out waiting for the condition
  Warning  FailedMount  14m (x3 over 23m)     kubelet, org2      Unable to attach or mount volumes: unmounted volumes=[hlf-channels orderer-tlsca peer-karga-peer9-tls peer-karga-peer9-admin-msp], unattached volumes=[hlf-channels default-token-prx47 podmetadata docker-sock my-minio-cred orderer-tlsca peer-karga-peer9-tls peer-karga-peer9-admin-msp]: timed out waiting for the condition
  Warning  FailedMount  8m19s (x21 over 34m)  kubelet, org2      MountVolume.SetUp failed for volume "orderer-tlsca" : secret "hlf-orderer--groeifabriek-tlsca" not found
  Warning  FailedMount  4m14s (x23 over 34m)  kubelet, org2      MountVolume.SetUp failed for volume "hlf-channels" : secret "hlf-channels" not found

```

**Comment by raftAtGit** [comment]
you need to run argo flow in the same namespace you created fabric components. the reason is, it needs to mount relevant `Secrets` and `ConfigMaps`. 

this is visible at the bottom of your output:
```
MountVolume.SetUp failed for volume "orderer-tlsca" : secret "hlf-orderer--groeifabriek-tlsca" not found
etc.
```

did you manage to run argo sample in the default namespace?

**Comment by weishancc** [comment]
> you need to run argo flow in the same namespace you created fabric components. the reason is, it needs to mount relevant `Secrets` and `ConfigMaps`.
> 
> this is visible at the bottom of your output:
> 
> ```
> MountVolume.SetUp failed for volume "orderer-tlsca" : secret "hlf-orderer--groeifabriek-tlsca" not found
> etc.
> ```
> 
> did you manage to run argo sample in the default namespace?

I ran the `hello world` and `coinflip argo` sample, both of them **didn't work** with default ns (e.g. argo submit --watch coinflip.yaml) and only worked with argo ns.

**Comment by raftAtGit** [comment]
> I ran the hello world and coinflip argo sample, both of them didn't work with default ns (e.g. argo submit --watch coinflip.yaml) and only worked with argo ns.

what was the error exactly? stuck in pending state or?

**Comment by weishancc** [comment]
> > I ran the hello world and coinflip argo sample, both of them didn't work with default ns (e.g. argo submit --watch coinflip.yaml) and only worked with argo ns.
> 
> what was the error exactly? stuck in pending state or?

Yes stck in pending state. Here's the log of coinflip workflow if necessary:

**k describe workflow coinflip-zsgk5**
```console
Name:         coinflip-zsgk5
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  argoproj.io/v1alpha1
Kind:         Workflow
Metadata:
  Creation Timestamp:  2020-10-12T09:42:26Z
  Generate Name:       coinflip-
  Generation:          1
  Resource Version:    60116
  Self Link:           /apis/argoproj.io/v1alpha1/namespaces/default/workflows/coinflip-zsgk5
  UID:                 498bd869-119d-4511-a89a-b4a09e2933d5
Spec:
  Arguments:
  Entrypoint:  coinflip
  Templates:
    Arguments:
    Inputs:
    Metadata:
    Name:  coinflip
    Outputs:
    Steps:
      [map[arguments:map[] name:flip-coin template:flip-coin]]
      [map[arguments:map[] name:heads template:heads when:{{steps.flip-coin.outputs.result}} == heads] map[arguments:map[] name:tails template:tails when:{{steps.flip-coin.outputs.result}} == tails]]
    Arguments:
    Inputs:
    Metadata:
    Name:  flip-coin
    Outputs:
    Script:
      Command:
        python
      Image:  python:alpine3.6
      Name:
      Resources:
      Source:  import random
result = "heads" if random.randint(0,1) == 0 else "tails"
print(result)

    Arguments:
    Container:
      Args:
        echo "it was heads"
      Command:
        sh
        -c
      Image:  alpine:3.6
      Name:
      Resources:
    Inputs:
    Metadata:
    Name:  heads
    Outputs:
    Arguments:
    Container:
      Args:
        echo "it was tails"
      Command:
        sh
        -c
      Image:  alpine:3.6
      Name:
      Resources:
    Inputs:
    Metadata:
    Name:  tails
    Outputs:
Status:
  Finished At:  <nil>
  Started At:   <nil>
Events:         <none>

```

**Comment by raftAtGit** [comment]
himm, run this command and try running argo sample in the default namespace again please
```
kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=default:default
```

normally, without this you got a different error message, but you never know

**Comment by weishancc** [comment]
I've already tried this role binding too, but thanks!
Just realize that argo patched arm version after [v2.11.0-rc1 ](https://github.com/argoproj/argo/releases/tag/v2.11.0-rc1) thus i'll try diff argo version.

**Comment by weishancc** [comment]
Hi, Raft : )

After changing argo version, now almost done !

Just one more question, where can i change fabric-baseimage? The screenshot is the log of peer when instantiating chaincodes, thanks.

![image](https://user-images.githubusercontent.com/44445847/95809610-c68b8600-0d41-11eb-9281-9ceface7a520.png)

**Comment by raftAtGit** [comment]
cool, I'm glad you made progress :)

are you using 0.7 branch? 

not sure what base image is, but I'm guessing it might be similar to [this issue](https://github.com/hyfen-nl/PIVT/issues/71). [the  fix](https://github.com/APGGroeiFabriek/PIVT/commit/6bbb72937b34b5c9f78dd61599e4c6a4cbdacd5d) mentioned there is not applied to master branch.

try applying something similar

**Comment by weishancc** [comment]
Sorry for replying late, so busy recently !

I re-tag baseimage and it could be find, whereas in instantiate process showed timeout error(maybe was the problem of baseimage, im still solving it, and this image is use to run node chaincode)

Alternately, i want to try golang chaincode. Since golang cannot directly run in peer container as node, i changed `fabric-peer `image to `fabric-tools`. After that, the log of peer showed 
```console 
Error: error getting chaincode code sf-iot: path to chaincode does not exist: /opt/gopath/src/chaincode/sf-iot
```
where sf-iot is my go chaincdoe, and this path is the original official volume path to chaincode(in this repo is just `/chaincode`) have you ever tested golang chaincode and occurred the error? thanks :)

**Comment by raftAtGit** [comment]
I dont have much experience with Golang chaincode, but it might be related to this, cherry-pick [this commit](https://github.com/hyfen-nl/PIVT/commit/1c8c599519ae73cdc60bece6bfa431890db7608e) please

**Comment by weishancc** [comment]
Hi Raft, 

I used master and upgrade my HLF version to v1.4.4 in the end but stuck when invoking chaincode(even-simpler).  

The chaincode instantiated with no error and chaincode container started up, whereas it cannot find instantiated chaincode.

 In my experience, this error may occurred from fabric images or didn't clean enviroment (In both case i've test images on official fabric-samples and cleaned all env) Do you have any idea?  Thanks!

```console
Error: endorsement failure during invoke. response: status:500 message:"make sure the chaincode even-simpler has been successfully instantiated and try again: chaincode even-simpler not found"
```

**Comment by raftAtGit** [comment]
weird, no idea indeed.

what peer reports when calling `peer chaincode list --instantiated -C common` command?

**Comment by weishancc** [comment]
list command shows `even-simpler` now : )
Just my stupid fault that modified capabilities in `configtx.yaml` . . 

Above all, it works on RPi-4 finally! 
Thanks!

**Comment by weishancc** [comment]
Hi, @raftAtGit 

Long time no see, hope u r well : )

I ran [Adding new peer organizations](https://github.com/hyfen-nl/PIVT#adding-new-peer-organizations) and got the following error within all pods, is it related to arm arch?
```console
error while running "VolumeBinding" filter plugin for pod "hlf-orderer--sgorderer--orderer5-0": pod has unbound immediate PersistentVolumeClaims
```

**Comment by raftAtGit** [comment]
thanks @weishancc I hope you too :)

dont know if this is somehow related to arm arch. 

but I know this shouldnt be related to adding new peer organizations. error says it cannot for some reason mount necessary volumes (PersistentVolumeClaims), so this should also happen when first launching the network. 

make a `kubectl describe` for the pods and then for missing `PersistentVolumeClaims`. possibly they stuck at `pending` state.

the reason might be as simple as you are using wrong `StorageClass`.
```
kubectl get StorageClass
```
this should match [used StorageClass](https://github.com/hyfen-nl/PIVT/blob/master/fabric-kube/hlf-kube/values.yaml#L22)

**Comment by weishancc** [comment]
I adopted  [nfs provisioner](https://github.com/helm/charts/tree/master/stable/nfs-client-provisioner) as storageclass. But it's so weird that the error still existed. I found **couchdb** container always CrashLoopBackOff where peer container was running well in peer pod. Also, command  `k logs hlf-peer--sgorg1--peer0-0 couchdb` showed nothing. Do u have any idea? Some maybe useful information is attached. Thanks!

![tempsnip](https://user-images.githubusercontent.com/44445847/101144753-c3cb5400-3653-11eb-88be-428b440cded1.png)
